Add a cache to reduce DB operations

somniumism opened the pull request:
I implemented the cache in our Merkle trie using the `lru_cache` crate that used in CodeChain.

I found that performance improved by up to 30 percent when reading the value from the trie. However, when writing the value to the trie, there was no change in performance.

The key(address)-value(rlp_node) is stored in the cache when they are written, read in the DB. Before DB operation(`get()`), we can find the value in the cache. Therefore, the overhead derived from DB operations will be reduced by adding cache.

I used `RefCell` because of an error `'self' is a '&' reference, so the data it refers to cannot be borrowed as mutable`

Commit:
Add a cache to reduce DB operations

The address-value pair is stored in the cache when the are written and read in the DB. As cache added, we can find the value in the cache. Therefore, the overhead derived from DB operations will be reduced.

somniumism commented on Cargo.toml:
```diff
@@ -14,6 +14,7 @@ rand = "0.6.1"
 rlp = { git = "https://github.com/CodeChain-io/rlp.git", version = "0.4" }
 rlp_derive = { git = "https://github.com/CodeChain-io/rlp.git", version = "0.2" }
 snap = "0.2"
+lru-cache = "0.1.2"
```
CodeChain already uses this crate.

somniumism commented on src/triedb.rs:
```diff
@@ -19,7 +19,9 @@ use crate::node::Node as RlpNode;
 use cdb::HashDB;
+use lru_cache::LruCache;
 use primitives::H256;
+use std::cell::RefCell;
```
I used `RefCell` because of an error `'self' is a '&' reference, so the data it refers to cannot be borrowed as mutable`

somniumism commented on src/triedb.rs:
```diff
@@ -52,12 +55,14 @@ impl<'db> TrieDB<'db> {
     /// Create a new trie with the backing database `db` and `root`
     /// Returns an error if `root` does not exist
     pub fn try_new(db: &'db dyn HashDB, root: &'db H256) -> crate::Result<Self> {
+        let cache: RefCell<LruCache<H256, Vec<u8>>> = RefCell::new(LruCache::new(3000));
```
It is an initialization of the cache. In this commit, the cache size is 3000. That means we can store 3,000 key-value pairs.

somniumism commented on src/triedb.rs:
```diff
@@ -71,9 +76,17 @@ impl<'db> TrieDB<'db> {
-                match RlpNode::decoded(&node_rlp) {
+                let node_rlp;
+                let decoded_rlp = if self.cache.borrow_mut().contains_key(&hash) {
+                    node_rlp = self.cache.borrow_mut().get_mut(&hash).unwrap().to_vec();
```
`self.cache.borrow_mut().get_mut(&hash).unwrap().to_vec()` and `self.cache.borrow_mut().insert(hash, (&*node_rlp).to_vec())` have different return types. 

Thus, I solved the two types of conflict through `decoded_rlp`. I need your advice on this. Is it okay to implement it like this?

sgkim126 replied:
```rust
                let mut cache = self.cache.borrow_mut();
                let cached_node = cache.get_mut(&hash);
                let node_rlp;
                let decoded_rlp = match cached_node {
                    Some(node_rlp) => {
                        RlpNode::decoded(node_rlp)
                    }
                    None => {
                        node_rlp = self.db.get(&hash).ok_or_else(|| TrieError::IncompleteDatabase(hash))?;
                        self.cache.borrow_mut().insert(hash, (&*node_rlp).to_vec());
                        RlpNode::decoded(&node_rlp)
                    }
                };

```
How about it?
It's a little more complicated, but the current code queries the cache twice when the data is cached.

somniumism replied:
It's good! I needed this feedback. It seems less complicated than I do. I'll modify it right away.

somniumism replied:
@sgkim126 It is a very good point, but It seems difficult to implement it as it is. This function uses recursion, then an error derives from `let mut cache = self.cache.borrow_mut();` because `borrow_mut()` duplicates. I tried to modify to solve the error, but I couldn't find a better idea.

sgkim126 replied:
Okay. How about leaving a fixme comment?

somniumism replied:
Good idea!

somniumism replied:
Modified : )

somniumism commented on src/triedbmut.rs:
```diff
@@ -20,6 +20,7 @@ use crate::triedb::TrieDB;
 use crate::{Trie, TrieError, TrieMut};
 use ccrypto::{blake256, BLAKE_NULL_RLP};
 use cdb::{DBValue, HashDB};
+use lru_cache::LruCache;
```
There is no change in performance when writing the value to the trie, so I will exclude the cache from this.

somniumism commented on src/triedbmut.rs:
```diff
@@ -31,16 +32,18 @@ pub(crate) struct TrieDBMut<'a> {
     pub fn new(db: &'a mut dyn HashDB, root: &'a mut H256) -> Self {
         *root = BLAKE_NULL_RLP;
-
+        let cache: LruCache<H256, Vec<u8>> = LruCache::new(3000);
```
There is no change in performance when writing the value to the trie, so I will exclude the cache from this.

somniumism commented on src/triedbmut.rs:
```diff
@@ -67,16 +71,24 @@ impl<'a> TrieDBMut<'a> {
-                let node_rlp = self.db.get(&hash).ok_or_else(|| TrieError::IncompleteDatabase(hash))?;
-
-                match RlpNode::decoded(&node_rlp) {
+                let node_rlp;
```
There is no change in performance when writing the value to the trie, so I will exclude the cache from this.

sgkim126 replied:
A write-through cache doesn't increase writing performance, but it improves reading performance.
Please add the test case that reads data after inserting data.
It will show how much the write-through cache has improved writing performance.

sgkim126 commented:
Why did you cache the address-node pair?
How about caching the key-value pair?
It will reduce the cache size and improve performance.

somniumism commented:
@sgkim126 I replaced the address-node pair type to key-value pair type and added some benchmark tests. I didn't test its performance yet. So, after testing, I will upload the result.

somniumism commented:
@sgkim126 @majecty I modified the source to cache `key-value pair type` here: https://github.com/CodeChain-io/rust-merkle-trie/pull/7/commits/dc05eba8c280054551d5904e2115152701997026

And I added source to test performances here: https://github.com/CodeChain-io/rust-merkle-trie/pull/7/commits/6ef5305b62c166762bf376bb04c125b2add53f1f

---

I tested control(no change), `address-node pair type`, and `key-value pair type`.
At first, I compared the performance between `address-node pair` and `key-value pair`.
In a condition that cache size is 5000, the results are here:

Address-node pair type:
```
test bench_3read_per_10write ... bench: 274,078,254 ns/iter (+/- 42,229,817)
test bench_read_after_write  ... bench: 274,222,428 ns/iter (+/- 33,002,606)
test bench_read_multiple     ... bench: 109,009,176 ns/iter (+/- 4,381,515)
test bench_read_single       ... bench:      20,080 ns/iter (+/- 2,642)
test bench_write_multiple    ... bench: 243,785,129 ns/iter (+/- 27,583,887)
test bench_write_single      ... bench:       8,251 ns/iter (+/- 293)
```

Key-value pair type(new implemented):
```
test bench_3read_per_10write ... bench: 262,793,359 ns/iter (+/- 33,333,621)
test bench_read_after_write  ... bench: 265,758,796 ns/iter (+/- 35,558,533)
test bench_read_multiple     ... bench: 171,219,118 ns/iter (+/- 4,973,717)
test bench_read_single       ... bench:      18,736 ns/iter (+/- 1,640)
test bench_write_multiple    ... bench: 239,363,551 ns/iter (+/- 37,571,733)
test bench_write_single      ... bench:       8,550 ns/iter (+/- 268)
```

Result: The performance of `address-node pair type` is better than `key-value pair type` in `betch_read_multiple`. In new test(`bench_3read_per_10write` and `bench_read_after_write`),  There was no big difference.

---

At second, At first, I compared the performance of three sources (control, add-node, and key-value) In a condition that cache size is 3000, the results are here:

Control(No change):
```
test bench_3read_per_10write ... bench: 241,656,430 ns/iter (+/- 27,254,352)
test bench_read_after_write  ... bench: 243,835,078 ns/iter (+/- 31,713,887)
test bench_read_multiple     ... bench: 164,791,657 ns/iter (+/- 4,273,903)
test bench_read_single       ... bench:      17,803 ns/iter (+/- 1,791)
test bench_write_multiple    ... bench: 220,595,707 ns/iter (+/- 27,850,139)
test bench_write_single      ... bench:       7,329 ns/iter (+/- 508)
```

Address-node pair type:
```
test bench_3read_per_10write ... bench: 273,057,890 ns/iter (+/- 35,250,910)
test bench_read_after_write  ... bench: 273,548,956 ns/iter (+/- 31,988,101)
test bench_read_multiple     ... bench: 115,298,526 ns/iter (+/- 2,858,011)
test bench_read_single       ... bench:      19,995 ns/iter (+/- 1,786)
test bench_write_multiple    ... bench: 245,025,401 ns/iter (+/- 47,407,288)
test bench_write_single      ... bench:       9,264 ns/iter (+/- 2,692)
```

Key-value pair type(new implemented):
```
test bench_3read_per_10write ... bench: 274,508,661 ns/iter (+/- 139,046,294)
test bench_read_after_write  ... bench: 279,826,302 ns/iter (+/- 109,041,533)
test bench_read_multiple     ... bench: 137,021,858 ns/iter (+/- 8,952,564)
test bench_read_single       ... bench:      18,988 ns/iter (+/- 1,451)
test bench_write_multiple    ... bench: 246,653,602 ns/iter (+/- 61,970,458)
test bench_write_single      ... bench:       8,901 ns/iter (+/- 733)
```

Result: The performance of `address-node pair type` is better than `key-value pair type` and `control` in`betch_read_multiple`. In the new test(`bench_3read_per_10write` and `bench_read_after_write`), the performance of control is better than both two-source with the cache. This means that the implementation of the cache had no effect.

---

I'm shocked because I can't get the results I want. I can't believe that adding A had no effect on performance.
Please look at the `key-value pair type` cache and new benchmark test I implemented, and let me know if there is any problem. I hope my implementation(the new cache and test) is wrong.

sgkim126 commented:
Fundamentally, Merkle Trie is key-value storage with the Merkle proof. So what I expected when I said caching key-value pairs is modifying [`get`](https://github.com/CodeChain-io/rust-merkle-trie/blob/a4400e98af6241f5da4798025d5b13e39e45b1a7/src/triedbmut.rs#L414), [`insert`](https://github.com/CodeChain-io/rust-merkle-trie/blob/a4400e98af6241f5da4798025d5b13e39e45b1a7/src/triedbmut.rs#L426), and [`remove`](https://github.com/CodeChain-io/rust-merkle-trie/blob/a4400e98af6241f5da4798025d5b13e39e45b1a7/src/triedbmut.rs#L435), the top-level interfaces which directly use a key instead of a path.

By the way, I think we should enhance the benchmark tests. The current tests seem to inquiry about random keys, so they might not hit the cache. I cannot assert, but the recently updated and read keys may be used soon again. It's the assumption that we want to use the cache. I think the benchmark tests following this assumption would be needed.

sgkim126 commented:
@somniumism and I discussed this issue. It seems that benchmarking real-like scenarios would take a long time. So we will make it caches the trie node in this commit and update the caching policy after implementing the benchmark.

sgkim126 commented on src/triedbmut.rs:
```diff
@@ -67,16 +73,25 @@ impl<'a> TrieDBMut<'a> {
-                let node_rlp = self.db.get(&hash).ok_or_else(|| TrieError::IncompleteDatabase(hash))?;
-
-                match RlpNode::decoded(&node_rlp) {
+                let node_rlp;
```
ditto
